server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/SPFTest.java

```diff
@@ -0,0 +1,238 @@
+        Mailet mailet = testMailet();
+
+        mailet.service(mail);
+        assertThat(AttributeUtils.getValueAndCastFromMail(mail, RESULT_ATTRIBUTE, String.class)
+            .orElse(null)).isEqualTo("permerror");
+    }
+
+    @Test
+    public void serviceShouldPerformSPFCheckWithResultTempError() throws MessagingException {
+        FakeMail mail = fakeMail().sender("hello@spf5.james.apache.org").build();
+        Mailet mailet = testMailet();
+
+        mailet.service(mail);
+        assertThat(AttributeUtils.getValueAndCastFromMail(mail, RESULT_ATTRIBUTE, String.class)
+            .orElse(null)).isEqualTo("temperror");
+    }
+
+    private Mailet testMailet() throws MessagingException {
+        return testMailet(false, false, "127.0.0.0/8");
+    }
+
+    private Mailet testMailet(boolean debug, boolean addHeader, String ignoreNetworks)
+        throws MessagingException {
+        FakeMailetConfig mailetConfig = FakeMailetConfig.builder()
+            .mailetName(MAILET_NAME)
+            .mailetContext(FakeMailContext.defaultContext())
+            .setProperty("debug", String.valueOf(debug))
+            .setProperty("addHeader", String.valueOf(addHeader))
+            .setProperty("ignoreNetworks", ignoreNetworks)
+            .build();
+
+        SPF mailet = new SPF(mockedDnsService());
+        mailet.setSPFDnsService(mockedSPFDnsService());
+        mailet.init(mailetConfig);
+        return mailet;
+    }
+
+    private FakeMail.Builder fakeMail() {
+        return FakeMail.builder()
+            .name(MAILET_NAME)
+            .sender(MailAddress.nullSender())
+            .recipient(MailAddressFixture.ANY_AT_JAMES)
+            .remoteHost("some.host.local")
+            .remoteAddr("10.11.12.13");
+    }
+
+    private DNSService mockedDnsService() {
+        return new DNSService() {
+
+            @Override
+            public Collection<String> findMXRecords(String hostname) {
+                throw new UnsupportedOperationException("not supported");
+            }
+
+            @Override
+            public Collection<String> findTXTRecords(String hostname) {
+                throw new UnsupportedOperationException("not supported");
+            }
+
+            @Override
+            public Collection<InetAddress> getAllByName(String host) {
+                throw new UnsupportedOperationException("not supported");
+            }
+
+            @Override
+            public InetAddress getByName(String host) throws UnknownHostException {
+                return InetAddress.getByName(host);
+            }
+
+            @Override
+            public InetAddress getLocalHost() {
+                throw new UnsupportedOperationException("not supported");
+            }
+
+            @Override
+            public String getHostName(InetAddress addr) {
+                throw new UnsupportedOperationException("not supported");
+            }
+        };
+    }
```

Comment:
Using mockito maybe?

```
DNSServer mockDNSService = mock(DNSService.class);
when(mockDNSService.getByName(any())).thenAnswer(...);
return mockDNSService;
```

This significantly reduces the boiler plate thus here should be favored.